src/main/java/edu/umich/lib/dor/replicaexperiment/service/Package.java

```diff
@@ -18,6 +20,14 @@ public Path getRootPath() {
         return depositDir.resolve(packagePath);
     }
 
+    public void validatePath() {
```

Comment:
I'll take a look at this tomorrow, what you said makes sense.

Reply:
@botimer, if you have a moment, please take a look at this commit: https://github.com/mlibrary/dor-replica-experiment/pull/16/commits/361c347de2b62481eefe22b1f71da752e34cbd52

I think I made reasonable choices, but I was a little uncertain about the `DepositDirectoryTest` change. Basically, having `validatePath` in the `Package` constructor means that it will get called in `DepositDirectory.getPackage`. I elected to just make `DepositDirectory` depend on a fixture like `Package`, but there might be another option.

Reply:
Yeah, I think it's great. What we see here is that details about `Package`'s responsibilities falling away from the other class's tests. The tests are giving you feedback about the design change -- this was originally overspecification, and now that rule (directory exists) is enforced in `Package`... So we don't mention it from outside, but rather trust that `Package` is doing its job, and specify the behavior of `DepositDirectory` more narrowly.

Those two tests and the bit of mocking dropping away while retaining full behavioral coverage is a confirmation of the design change. The code has gotten simpler (less coupled) while still doing the same thing indicates: this is a good change!

Reply:
Now, to introduce a hypothetical complication:

If we found that we wanted to preemptively declare a Package -- something prototypical or to-be-created -- this design would "fight" us. That's not necessarily a bad thing, because it would put the focus on the exact responsibilities of what a Package is supposed to be doing. As it stands, it corresponds to "some existing stuff on disk". If we need something else, the question is not about relaxing that constraint, but asking _what else_ we need.

It may be that we have another kind of prototype that can be realized as a Package... Or it may be that we have a logical, or in-memory variant of a Package... Lots of modeling options that would depend on the exact circumstances. But the main point is: as we have defined it, the Package makes sense. It enforces the business rules that we want enforced. If we need _another kind of thing_, we should introduce it as such.

It's already easy enough to use this Package for testing purposes. We could use a real one with real fixtures on disk, or we can use a fake/mock one that responds with non-disk data. I see no problem with the validation being enforced in the constructor. This actually simplifies the logical variation that can happen, driving out opportunities for bugs. Having your main objects able to optionally ignore their business rules (i.e., their reason for existing) is an invitation for weird behavior, so we just plain avoid that.